### PR TITLE
Add Builder to help create Transactions easier

### PR DESCRIPTION
**Pull request by achiurizo**
This adds a easy way to craft Transctions. Currently, you have to build the entire Transaction structs by tediously, now you can:

```elixir
new()
|> add_input(blknum: 1, txindex: 0, oindex: 0)
|> add_output(output_guard: <<1::160>>, token: <<0::160>>, amount: 1)
```

Additionally, this cleans up some of the typespecs and enforces structs for the underlying data.

### CODE REVIEW

File: lib/ex_plasma/output.ex [hunk 1]
```diff
@@ -170,20 +174,20 @@ defmodule ExPlasma.Output do
 
   # Validate the output ID. Bypass the validation if it doesn't
   # exist in the output body.
-  defp do_validate_id(%{output_id: nil} = output), do: {:ok, output}
+  defp do_validate_id(%__MODULE__{output_id: nil} = output), do: {:ok, output}
 
-  defp do_validate_id(%{output_id: %{} = output_id}),
-    do: ExPlasma.Output.Position.validate(output_id)
+  defp do_validate_id(%__MODULE__{output_id: %{} = output_id}),
```
Comment (InoMurko) [inline lib/ex_plasma/output.ex:179]:
```suggestion
  defp do_validate_id(%__MODULE__{output_id: %{} = output_id}), do: Position.validate(output_id)
```

if you want to do this, maybe it's worth adjusting the formatter and credo by adding 
`  line_length: 120,`

` {Credo.Check.Readability.MaxLineLength, [priority: :low, max_length: 120]},`
same as in childchain!

Comment (InoMurko) [reply lib/ex_plasma/output.ex:179]: maybe it already works and you need to convince the formatter

Comment (achiurizo) [reply lib/ex_plasma/output.ex]: Already have it: https://github.com/omisego/ex_plasma/blob/master/.credo.exs#L99

File: test/ex_plasma/builder_test.exs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  test "builds a transaction with both inputs and outputs" do
+    txn =
+      [tx_type: 1, metadata: <<1::160>>, tx_data: 0]
+      |> new()
+      |> add_input(blknum: 1, txindex: 0, oindex: 0)
+      |> add_input(blknum: 2, txindex: 1, oindex: 0)
+      |> add_input(blknum: 3, txindex: 0, oindex: 1)
+      |> add_output(
+        output_type: 1,
+        output_data: %{output_guard: <<1::160>>, token: <<0::160>>, amount: 1}
```
Comment (InoMurko) [inline test/ex_plasma/builder_test.exs:16]:
```suggestion
      |> add_input(blknum: 3, txindex: 0, oindex: 1)
      |> add_output(
        output_type: 1,
        output_data: %{output_guard: <<1::160>>, token: <<0::160>>, amount: 1}
```
whoa!

**Pull request merged**